extensions/binding/org.eclipse.smarthome.binding.bluetooth/src/main/java/org/eclipse/smarthome/binding/bluetooth/BluetoothCharacteristic.java

```diff
@@ -0,0 +1,633 @@
+    }
+
+    /**
+     * Returns a descriptor with a given UUID out of the list of
+     * descriptors for this characteristic.
+     *
+     * @return the {@link BluetoothDescriptor}
+     */
+    public BluetoothDescriptor getDescriptor(UUID uuid) {
+        return gattDescriptors.get(uuid);
+    }
+
+    /**
+     * Get the stored value for this characteristic.
+     *
+     */
+    public int[] getValue() {
+        return value;
+    }
+
+    /**
+     * Get the stored value for this characteristic.
+     *
+     */
+    public byte[] getByteValue() {
+        byte[] byteValue = new byte[value.length];
+        for (int cnt = 0; cnt < value.length; cnt++) {
+            byteValue[cnt] = (byte) (value[cnt] & 0xFF);
+        }
+        return byteValue;
+    }
+
+    /**
+     * Return the stored value of this characteristic.
+     *
+     */
+    public Integer getIntValue(int formatType, int offset) {
+        if ((offset + getTypeLen(formatType)) > value.length) {
+            return null;
+        }
+
+        switch (formatType) {
+            case FORMAT_UINT8:
+                return unsignedByteToInt(value[offset]);
+
+            case FORMAT_UINT16:
+                return unsignedBytesToInt(value[offset], value[offset + 1]);
+
+            case FORMAT_UINT32:
+                return unsignedBytesToInt(value[offset], value[offset + 1], value[offset + 2], value[offset + 3]);
+
+            case FORMAT_SINT8:
+                return unsignedToSigned(unsignedByteToInt(value[offset]), 8);
+
+            case FORMAT_SINT16:
+                return unsignedToSigned(unsignedBytesToInt(value[offset], value[offset + 1]), 16);
+
+            case FORMAT_SINT32:
+                return unsignedToSigned(
+                        unsignedBytesToInt(value[offset], value[offset + 1], value[offset + 2], value[offset + 3]), 32);
+        }
+
+        return null;
+    }
+
+    /**
+     * Return the stored value of this characteristic. This doesn't read the remote data.
+     *
+     */
+    public Float getFloatValue(int formatType, int offset) {
+        if ((offset + getTypeLen(formatType)) > value.length) {
+            return null;
+        }
+
+        switch (formatType) {
+            case FORMAT_SFLOAT:
+                return bytesToFloat(value[offset], value[offset + 1]);
+
+            case FORMAT_FLOAT:
+                return bytesToFloat(value[offset], value[offset + 1], value[offset + 2], value[offset + 3]);
+        }
+
+        return null;
+    }
+
+    /**
+     * Return the stored value of this characteristic. This doesn't read the remote data.
+     *
+     */
+    public String getStringValue(int offset) {
+        if (value == null || offset > value.length) {
+            return null;
+        }
+        byte[] strBytes = new byte[value.length - offset];
+        for (int i = 0; i != (value.length - offset); ++i) {
+            strBytes[i] = (byte) value[offset + i];
+        }
+        return new String(strBytes);
```

Comment:
`new String(byte[])` is the "I'm feeling lucky"-version wrt the character encoding. Rather give it _some_  fixed charset like UTF-8 instead of relying on the platform default.